src/lib/proof-system/zkfunction.ts

```diff
@@ -0,0 +1,185 @@
+      const publicInputSize = publicInputType.sizeInFields();
+      const publicInputFields = publicInputType.toFields(publicInput);
+      const main = mainFromCircuitData(config, privateInputs);
+      await initializeBindings();
+      return withThreadPool(async () => {
+        const proof = Snarky.circuit.prove(
+          main,
+          publicInputSize,
+          MlFieldConstArray.to(publicInputFields),
+          _keypair!.value
+        );
+        return new Proof(proof);
+      });
+    },
+
+    /**
+     * Verifies a proof using the public input, the proof, and optionally a verification key of the circuit(ZkFunction).
+     *
+     * @param publicInput The public input to the circuit.
+     * @param proof The proof to verify.
+     * @returns `true` if the proof is valid, otherwise `false`.
+     *
+     * @throws If `compile` has not been called.
+     *
+     * @example
+     * ```ts
+     * const verificationKey = await zkf.compile();
+     * const proof = await zkf.prove([privateInput], publicInput);
+     * const isValid = await zkf.verify(publicInput, proof, verificationKey);
+     * ```
+     */
+    async verify(publicInput: P, proof: Proof, verificationKey?: VerificationKey) {
+      verificationKey = verificationKey ?? _keypair!.verificationKey();
+      if (!_keypair) throw new Error('Cannot find VerificationKey. Please call compile() first!');
+      const publicInputFields = publicInputType.toFields(publicInput);
+      await initializeBindings();
+      return prettifyStacktracePromise(
+        withThreadPool(async () =>
+          Snarky.circuit.verify(
+            MlFieldConstArray.to(publicInputFields),
+            proof.value,
+            verificationKey.value
+          )
+        )
+      );
+    },
+  };
+}
+
+class Keypair {
+  value: Snarky.Keypair;
+
+  constructor(value: Snarky.Keypair) {
+    this.value = value;
+  }
+
+  verificationKey() {
+    return new VerificationKey(Snarky.circuit.keypair.getVerificationKey(this.value));
+  }
+
+  /**
+   * Returns a low-level JSON representation of the {@link ZkFunction} from its {@link Keypair}:
+   * a list of gates, each of which represents a row in a table, with certain coefficients and wires to other (row, column) pairs
+   * @example
+   * ```ts
+   * const keypair = await zkf.compile();
+   * const json = MyProvable.witnessFromKeypair(keypair);
+   * ```
+   */
+  constraintSystem() {
+    try {
+      return gatesFromJson(Snarky.circuit.keypair.getConstraintSystemJSON(this.value)).gates;
+    } catch (error) {
+      throw prettifyStacktrace(error);
+    }
+  }
+}
+
+/**
+ * Proofs can be verified using a {@link VerificationKey} and the public input.
+ */
+class Proof {
+  value: Snarky.Proof;
+
+  constructor(value: Snarky.Proof) {
+    this.value = value;
+  }
+}
+
+/**
+ * Part of the circuit {@link Keypair}. A verification key can be used to verify a {@link Proof} when you provide the correct public input.
+ */
+class VerificationKey {
```

Comment:
Same here - we need to make sure that somehow we can differenciate between the different types of verification keys